### PR TITLE
Rustdoc migrate to table so the gui can handle >2k constants

### LINKED ISSUE #88545: Generated docs on long pages render text on top of itself
_I hope this is the right repo to report this, I wasn't sure if it should go in the main rust repo instead._

The generated docs for pages with a crazy number of items seem to give up on positioning and then just dump all subsequent entries on the same line past a certain point. It looks like it might be related to hardware acceleration, but I'm not sure.

For example, viewing https://doc.rust-lang.org/beta/core/arch/arm/index.html in Chrome 92 on Windows and hitting ctrl+pgdwn to go to the very bottom shows this:

image

### PR DESCRIPTION
**Pull request by dns2utf8**
Closes #88545.

This PR adds a test for overlapping entries in the `item-table` https://github.com/rust-lang/rust/issues/88545
It currently includes the commit with the workaround from https://github.com/rust-lang/rust/pull/88776

### CODE REVIEW

File: src/test/rustdoc-gui/huge-collection-of-constants.goml
Comment (GuillaumeGomez) [inline src/test/rustdoc-gui/huge-collection-of-constants.goml:3]:
```suggestion
// Make sure that the last two entries are more than 12 pixels apart and not stacked on each other.

```

File: src/test/rustdoc-gui/src/test_docs/build.rs
Comment (GuillaumeGomez) [inline src/test/rustdoc-gui/src/test_docs/build.rs:9]: 1200 should be enough I think. Since we rebuild doc every time, maybe better to not write too much.

File: src/test/rustdoc-gui/src/test_docs/build.rs
Comment (GuillaumeGomez) [inline src/test/rustdoc-gui/src/test_docs/build.rs:10]:
```suggestion
        output.push_str(&format!("/// Some const A{0}\npub const A{0}: isize = 0;\n", i);
```

File: src/test/rustdoc-gui/src/test_docs/build.rs
Comment (GuillaumeGomez) [inline src/test/rustdoc-gui/src/test_docs/build.rs:1]: To be updated if you change the number.

File: src/test/rustdoc-gui/src/test_docs/build.rs
Comment (GuillaumeGomez) [inline src/test/rustdoc-gui/src/test_docs/build.rs:15]:
```suggestion
    write(&[out_dir, "huge_amount_of_consts.rs"].iter().collect::<PathBuf>(), output)
    
```

**Comment by GuillaumeGomez** [comment]
Might be worth it to add a check on test and/or rustdoc hash version to check if we need to rebuild docs or not, even more with this change...

Or maybe we could just create this file once and for all. Not sure what's the best solution here. An opinion @Mark-Simulacrum ?

**Comment by GuillaumeGomez** [comment]
I marked this PR as blocked until #88776 is merged.

**Comment by dns2utf8** [comment]
> Might be worth it to add a check on test and/or rustdoc hash version to check if we need to rebuild docs or not, even more with this change...
Do you mean that we only run the tests that have failed or are new? Or about (re-)generating the HTML and such?

> Or maybe we could just create this file once and for all. Not sure what's the best solution here. An opinion @Mark-Simulacrum ?
We can store the generated file too, sure

File: src/test/rustdoc-gui/src/test_docs/build.rs
Comment (dns2utf8) [reply src/test/rustdoc-gui/src/test_docs/build.rs:9]: the grid-2 spec talks about -10k to +10k cells, it is currently a blink-engine thing to clamp at 1k. In some of their examples I saw them using 1.5k lines so I guessed 2k.
Switching to a fixed file instead of generating it all the time would reduce the amount of writing

Comment (GuillaumeGomez) [reply src/test/rustdoc-gui/src/test_docs/build.rs:9]: Then let's keep 2k and generate a file.

**Comment by dns2utf8** [comment]
I implemented both layouts with a table like structure but using `<div>`s and classes.
The file sizes of the `index.html` documents have increased by ~10% with this change, so maybe there is some potential there.

**Comment by GuillaumeGomez** [comment]
Oh wow, that's awesome! So just remains to generate a 2000 line file with the constants I guess for the GUI test? :)

Once done, let's merge it, great work, thanks!

**Comment by Mark-Simulacrum** [comment]
Please don't check in a 2000-line generated file, we can generate it at test runtime quickly enough that it shouldn't be necessary to pregenerate and bloat everyone's repository unnecessarily.

> the grid-2 spec talks about -10k to +10k cells, it is currently a blink-engine thing to clamp at 1k. In some of their examples I saw them using 1.5k lines so I guessed 2k.

I'd be interested in hearing more about our motivation for adding this test, though. Based on what's been said so far, it sounds like the test is intended to avoid us regressing back to using grids *specifically for this feature* -- but that seems like the wrong thing to target. In practice, if grids are problematic with large tables -- whether the clamp is at 1000 or 10,000 doesn't seem like it matters *that* much in practice - then it seems like the better approach is to add a tidy lint preventing their usage in our css files. Obviously, if we know ahead of time that the lint doesn't apply to some statically known small-size user, then we can avoid it in that case. But this would both save on the generation of a 2k line file and the related annoyances (e.g., test run time, etc.) and target the problem more holistically, right?

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum Then we need to improve how we run GUI tests because currently we regenerate test docs every time whether or not something was changed. I can do that in a follow-up PR.

Also, @dns2utf8 removed the grid and replaced it with something else, completely fixing the issue.

**Comment by Mark-Simulacrum** [comment]
> @Mark-Simulacrum Then we need to improve how we run GUI tests because currently we regenerate test docs every time whether or not something was changed. I can do that in a follow-up PR.

This seems to have nothing to do with my request? I expect each test to either run from scratch or not at all. We can add some caching atop that similar to compiletest, though.

> Also, @dns2utf8 removed the grid and replaced it with something else, completely fixing the issue.

Yes, I gathered that. What I'm saying is that testing that we can support 2,000 constants doesn't seem necessary, we can test directly for grid usage and lint against that.

**Comment by GuillaumeGomez** [comment]
> Yes, I gathered that. What I'm saying is that testing that we can support 2,000 constants doesn't seem necessary, we can test directly for grid usage and lint against that.

Then I don't understand what you mean.

**Comment by dns2utf8** [comment]
Currently, there is no use of grid layouts in rustdoc and adding a lint that prevents their use could work too for now.

If we decide to use a grid in the future for something else (like switching the layout between super large desktops and 1080p-like desktops) then the check might get removed again. Adding a comment to the lint could reduce that risk again.

Wild idea (maybe a bad one): what if we move some of the tests into a separate folder that is only run at release?

**Comment by Mark-Simulacrum** [comment]
No, we don't want a special category of tests, those just cause pain all around.

Yes, the lint would prevent usage of grids by accident; the reviewer/author could of course justify the usage in specific cases and silence the lint on that particular usage. Tidy lints don't just get dropped, so this would be a reliable way of ensuring we don't make this mistake in the future at very low overhead.

**Comment by GuillaumeGomez** [comment]
I'm very curious on how you want to write such a lint considering all the debate around https://github.com/rust-lang/rust/pull/86178 (which is still open)...

**Comment by Mark-Simulacrum** [comment]
A scan of css files in-tree (perhaps limited to rustdoc by directory) looking for "display: grid;" seems likely to be sufficient? This check is much easier to write and much less error prone, IMO; it need not scan a language with complex syntax or variables etc.

**Comment by GuillaumeGomez** [comment]
Oh right, I had in mind potential adds through inline CSS (very unlikely) and through JS. Wel, if we don't consider those two, it should be easy enough.

**Comment by dns2utf8** [comment]
A CSS lint seems reasonable.

If we want to keep the long check: what if we reexport some huge module (like one with the arm platform constants) in the test docs and use that for checking?
Or would you just remove this test for now because the table style is very stable I think?

**Comment by JohnCSimon** [comment]
Triage: 
@GuillaumeGomez 

> I marked this PR as blocked until #88776 is merged.

this has been merged

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Let's merge this for now then!

Thanks @dns2utf8 !

@bors: r+

**Comment by dns2utf8** [comment]
Thank you @Mark-Simulacrum for rebasing it! :blush:

**Pull request merged**
